File: packages/quick_actions/ios/Classes/QuickActionsPlugin.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "QuickActionsPlugin.h"

static NSString *const CHANNEL_NAME = @"plugins.flutter.io/quick_actions";

@interface FLTQuickActionsPlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@property (nonatomic, strong) NSString *shortcutType;
@end

@implementation FLTQuickActionsPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
  [FlutterMethodChannel methodChannelWithName:CHANNEL_NAME
                              binaryMessenger:[registrar messenger]];
  FLTQuickActionsPlugin *instance = [[FLTQuickActionsPlugin alloc] init];
  instance.channel = channel;
  [registrar addApplicationDelegate:instance];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  if ([call.method isEqualToString:@"setShortcutItems"]) {
    setShortcutItems(call.arguments);
    result(nil);
  } else if ([call.method isEqualToString:@"clearShortcutItems"]) {
    if (@available(iOS 9.0, *)) {
      [UIApplication sharedApplication].shortcutItems = @[];
    }
    result(nil);
  } else if ([call.method isEqualToString:@"getLaunchAction"]) {
    result(self.shortcutType); // This is used when the app is killed and open the first time via quick actions
    self.shortcutType = nil;
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (void)dealloc {
  [self.channel setMethodCallHandler:nil];
  self.channel = nil;
}

- (BOOL)application:(UIApplication *)application
WillFinishLaunchingWithOptions:(NSDictionary<UIApplicationLaunchOptionsKey, id> *)launchOptions {
  if (@available(iOS 9.0, *)) {
    UIApplicationShortcutItem *shortcutItem = launchOptions[UIApplicationLaunchOptionsShortcutItemKey];
    if(shortcutItem != NULL) {
      self.shortcutType = shortcutItem.type;
      [self.channel invokeMethod:@"launch" arguments:shortcutItem.type];
      return NO;
    } else {
      [self.channel invokeMethod:@"launch" arguments:nil];
      self.shortcutType = nil;
    }
  }
  return YES;
}

- (BOOL)application:(UIApplication *)application
performActionForShortcutItem:(UIApplicationShortcutItem *)shortcutItem
  completionHandler:(void (^)(BOOL succeeded))completionHandler API_AVAILABLE(ios(9.0)){
  self.shortcutType = shortcutItem.type;
  [self.channel invokeMethod:@"launch" arguments:shortcutItem.type];
  return YES;
}

#pragma mark Private functions

static void setShortcutItems(NSArray *items) {
  if (@available(iOS 9.0, *)) {
    NSMutableArray *newShortcuts = [[NSMutableArray alloc] init];
    for (id item in items) {
      UIApplicationShortcutItem *shortcut = deserializeShortcutItem(item);
      [newShortcuts addObject:shortcut];
    }
    [UIApplication sharedApplication].shortcutItems = newShortcuts;
  }
}

API_AVAILABLE(ios(9.0))
static UIApplicationShortcutItem *deserializeShortcutItem(NSDictionary *serialized) {
  UIApplicationShortcutIcon *icon =
  [serialized[@"icon"] isKindOfClass:[NSNull class]]
  ? nil
  : [UIApplicationShortcutIcon iconWithTemplateImageName:serialized[@"icon"]];
  
  return [[UIApplicationShortcutItem alloc] initWithType:serialized[@"type"]
                                          localizedTitle:serialized[@"localizedTitle"]
                                       localizedSubtitle:nil
                                                    icon:icon
                                                userInfo:nil];
}

@end

File: packages/quick_actions/quick_actions_ios/ios/Classes/QuickActionsPlugin.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Flutter

public final class QuickActionsPlugin: NSObject, FlutterPlugin {

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "plugins.flutter.io/quick_actions_ios",
      binaryMessenger: registrar.messenger())
    let instance = QuickActionsPlugin(channel: channel)
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
  }

  private let channel: FlutterMethodChannel
  private let shortcutStateManager: FLTShortcutStateManager
  /// The type of the shortcut item selected when launching the app.
  private var launchingShortcutType: String? = nil

  // TODO: (hellohuanlin) remove `@objc` attribute and make it non-public after migrating tests to Swift.
  @objc
  public init(
    channel: FlutterMethodChannel,
    shortcutStateManager: FLTShortcutStateManager = FLTShortcutStateManager()
  ) {
    self.channel = channel
    self.shortcutStateManager = shortcutStateManager
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setShortcutItems":
      // `arguments` must be an array of dictionaries
      let items = call.arguments as! [[String: Any]]
      shortcutStateManager.setShortcutItems(items)
      result(nil)
    case "clearShortcutItems":
      shortcutStateManager.setShortcutItems([])
      result(nil)
    case "getLaunchAction":
      result(nil)
    case _:
      result(FlutterMethodNotImplemented)
    }
  }

  public func application(
    _ application: UIApplication, performActionFor shortcutItem: UIApplicationShortcutItem,
    completionHandler: @escaping (Bool) -> Void
  ) -> Bool {
    handleShortcut(shortcutItem.type)
    return true
  }

  public func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [AnyHashable: Any] = [:]
  ) -> Bool {
    if let shortcutItem = launchOptions[UIApplication.LaunchOptionsKey.shortcutItem]
      as? UIApplicationShortcutItem
    {
      // Keep hold of the shortcut type and handle it in the
      // `applicationDidBecomeActive:` method once the Dart MethodChannel
      // is initialized.
      launchingShortcutType = shortcutItem.type

      // Return false to indicate we handled the quick action to ensure
      // the `application:performActionFor:` method is not called (as
      // per Apple's documentation:
      // https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622935-application).
      return false
    }
    return true
  }

  public func applicationDidBecomeActive(_ application: UIApplication) {
    if let shortcutType = launchingShortcutType {
      handleShortcut(shortcutType)
      launchingShortcutType = nil
    }
  }

  private func handleShortcut(_ shortcut: String) {
    channel.invokeMethod("launch", arguments: shortcut)
  }

}
